api/v1alpha1/dnsrecord_endpoints.go

```diff
@@ -0,0 +1,326 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	"github.com/kuadrant/dns-operator/internal/common/hash"
+)
+
+const (
+	SimpleRoutingStrategy       RoutingStrategy = "simple"
+	LoadBalancedRoutingStrategy RoutingStrategy = "loadbalanced"
+
+	DefaultTTL      = 60
+	DefaultCnameTTL = 300
+
+	ClusterIDLength = 6
+
+	LabelLBAttributeGeoCode = "kuadrant.io/lb-attribute-geo-code"
+)
+
+var (
+	ErrUnknownRoutingStrategy = fmt.Errorf("unknown routing strategy")
+)
+
+// RoutingStrategy specifies a strategy to be used: simple or load-balanced
+// +kubebuilder:validation:Enum=simple;loadbalanced
+// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="RoutingStrategy is immutable"
+// +kubebuilder:default=loadbalanced
+type RoutingStrategy string
+
+type CustomWeight struct {
+	Weight   int
+	Selector v1.LabelSelector
+}
+
+// Routing holds all necessary information to generate endpoints
+type Routing struct {
+	Strategy       RoutingStrategy
+	DefaultGeoCode string
+	DefaultWeight  int
+	CustomWeights  []CustomWeight
+	ClusterID      string
+}
+
+type RoutingBuilder struct {
+	*Routing
+}
+
+func NewRoutingBuilder() *RoutingBuilder {
+	return &RoutingBuilder{
+		Routing: &Routing{},
+	}
+}
+
+func (rb *RoutingBuilder) WithSimpleStrategy() *RoutingBuilder {
+	rb.Strategy = SimpleRoutingStrategy
+	return rb
+}
+
+func (rb *RoutingBuilder) WithLoadBalancedStrategy(clusterID, defaultGeo string, defaultWeight int) *RoutingBuilder {
+	rb.Strategy = LoadBalancedRoutingStrategy
+	rb.ClusterID = clusterID
+	rb.DefaultGeoCode = defaultGeo
+	rb.DefaultWeight = defaultWeight
+	return rb
+}
+
+func (rb *RoutingBuilder) WithCustomWeights(weights []CustomWeight) *RoutingBuilder {
+	rb.CustomWeights = weights
+	return rb
+}
+
+func (rb *RoutingBuilder) Build() (*Routing, error) {
+	return rb.Routing, rb.Validate()
+}
+
+func GenerateEndpoints(gateway *gatewayapiv1.Gateway, dnsRecord *DNSRecord, listener gatewayapiv1.Listener, routing *Routing) ([]*externaldns.Endpoint, error) {
```

Comment:
I'd prefer if we could do this without passing in the objects, e.g.
```
GenerateEndpoints(targets []string, host, strategy string) ([]*externaldns.Endpoint, error)
```

Reply:
that will be a lot of arguments: 

From Routing: 
- Strategy 
- Default Geo
- Default Weight 
- Custom weights 
- cluster ID 

From Gateway: 
- Addresses 
- Labels 

From DNS record: 
- name/namespace 

Listener: 
- hostname 

I "replaced" policy with `Routing` a struct and kept the rest, plus it allowed me to do a validation of the Routing args upon creation (ensuring we are passing correct info there). Why is there a preference towards more arguments than objects?

Reply:
if someone wants to use this to build DNS Records for the DNS Operator, but are not using the gateway API, the builder library is no use to them. Could we accept arguments both ways and have one function call the other internally?

Reply:
I understood that the idea was a builder like:
```
dnsBuilder
    .WithStrategy(x)
    .WithDefaultGeo(x)
    .WithClusterID(x)
    .WithAddress(x)
    .WithHostname(x)
    .Build()
```

Reply:
Also, why does it need the DNS Record name/namespace? The DNS Record might not exist yet.